### PR TITLE
MGDOBR-88: Update CD pipeline to build single docker images if a module has changed

### PR DESCRIPTION
**Pull request by r00ta**
[MGDOBR-88](https://issues.redhat.com/browse/MGDOBR-88) 

This PR aims to update the CD pipeline so to build a new docker image for each component, if it was modified in the merged pull request that triggered the action. 

The stategy is the following: if any of the "general" modules (`k8s`, `bom`, `infra`, etc...) has changed or if the single module has changed, then a new docker image for that component is built. If there are no general changes and no changes in the module, the docker image for that module is not built. 

Another possible strategy (that we do not have to maintain/update) is to build a docker image for every component regardless the changes: if the action was triggered, we re-build all the modules and all the docker images (and we update the demo environment as well).  

Depends on https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/81

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config)
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] All new functionality is tested and uses org.assertj.core.api.Assertions for Assertions
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: .github/workflows/CD.yml [hunk 1]
```diff
@@ -61,10 +65,69 @@ jobs:
...
         shell: bash
         run: |
-          docker build -f runner/src/main/docker/Dockerfile.jvm -t $OUTPUT_CONTAINER_NAME-jvm runner/
-          docker push $OUTPUT_CONTAINER_NAME-jvm
-          docker image rm -f $OUTPUT_CONTAINER_NAME-jvm
-#
+          echo "Building and pushing docker image for All-in-One"
+          docker build -f docker/Dockerfile.jvm -t $ALL_IN_ONE_CONTAINER_NAME-jvm runner/
+          docker push $ALL_IN_ONE_CONTAINER_NAME-jvm
+          docker image rm -f $ALL_IN_ONE_CONTAINER_NAME-jvm
```
Comment (robpblake) [inline .github/workflows/CD.yml:74]: @r00ta This `image rm -rf` means that the following steps to build a component have to do the entire `docker build` again. If we remove it, it means the following components should benefit from the Docker build cache and make the whole pipeline much quicker.

File: .github/workflows/CD.yml [hunk 2]
```diff
@@ -61,10 +65,69 @@ jobs:
...
+          docker build -f docker/Dockerfile.jvm -t $FLEET_MANAGER_CONTAINER_NAME-jvm runner/
+          docker push $FLEET_MANAGER_CONTAINER_NAME-jvm
+          docker image rm -f $FLEET_MANAGER_CONTAINER_NAME-jvm
+
+      # run only if 'ingress' or any 'general' modules have changed
+      - name: Build and Publish JVM Container - Fleet Shard
+        if: steps.filter.outputs.shard == 'true'
+        run: |
+          echo "Building and pushing docker image for Fleet Shard"
+          docker build -f docker/Dockerfile.jvm -t $FLEET_SHARD_CONTAINER_NAME-jvm runner/
```
Comment (robpblake) [inline .github/workflows/CD.yml:113]: @r00ta Do we need this `docker build` here? From what I see we're always building the same image no matter the component, and we always build the all-in-one first. We just need to retag (and push) the all-in-one image at this point.

File: .github/workflows/CD.yml (hunk 1 above)
Comment (r00ta) [reply .github/workflows/CD.yml]: Yeah it's a fair point. 
I've removed the ` docker image rm -f` from the steps and I've added a last clean up step. Wdyt @robpblake ?

Comment (robpblake) [reply .github/workflows/CD.yml]: @r00ta Just thinking about it..do we even need to do the clean-up of the container image at all? I assume the GH Action runner filesystem is "fresh" for each build?

**Review by robpblake — APPROVED** [review]
LGTM @r00ta . Cheers!

**Pull request merged**